### PR TITLE
First training iteration

### PR DESCRIPTION
**Pull request by IsmaPi**
@RaulPPelaez 

Fixed errors in the training loop of the models and trained both to have preliminary results. In the data generation (/src/data/generate.py)) I removed .flatten() calls that caused positions/forces to be passed to libMobility as (3N,) instead of (N,3). Also fixed the solver.Mdot() return value with proper tuple unpacking. Furthermore, I modified the evaluation pipeline (run.py) since there was a name mismatch. It now reads the strategy field from YAML content instead of deriving it from the config filename, so the evaluate stage correctly locates checkpoints. In the normalization (/src/data/normalization.py)) I fixed the zero-variance feature handling by setting std=1.0 for constant features (like theta), preventing division-by-near-zero explosion that was happening. 

For the training, I implemented early stopping (/src/training/trainer.py), (/src/utils/config.py) with a configurable patience (default 50 epochs) to prevent overfitting. Furthermore, I implemented theta normalization (/src/data/dataset.py) features in the normalization statistics and standardized them before being concatenated into node features.

For the config tuning, the GN model (/configs/model/torchmd_gn.yaml) size was reduced (hidden_dim 256→128, layers 4→3) to mitigate overfitting, the training configs (/configs/training/single_geometry.yaml), (/configs/training/mixed_geometry.yaml), (/configs/training/leave_one_out.yaml)) had their hyperparameters tuned to a lower learning rate, higher weight decay, longer warmup, tighter gradient clipping, and early stopping enabled.

New Files in the repo include (/scripts/visualize.py), which provides visualization utilities for results analysis; (/scripts/evaluate.py), which contains extended evaluation scripts with scaling benchmarks; and (/examples.ipynb), which is an example notebook showing how to implement libMobility properly in the code.

Results The Equivariant Transformer (ET) achieves 0.9% relative error on nbody_open, compared to 23.9% with the GN (SchNet), which is about a 26x improvement.

### CODE REVIEW

File: scripts/evaluate.py [hunk 1]
```diff
@@ -143,6 +152,54 @@ def main():
...
 
+    elif args.eval_mode == "rollout":
+        if args.data_config is None:
+            print("Error: --data-config required for rollout mode.")
+            sys.exit(1)
+
+        raw_data = load_yaml(args.data_config)
+        data_cfg = build_data_config(raw_data)
+
+        normalizer_path = "data/processed/normalizer.pt"
```
Comment (RaulPPelaez) [inline scripts/evaluate.py:163]: What is this normalizer?

File: scripts/train.py [hunk 2]
```diff
@@ -67,10 +65,9 @@ def build_datasets(data_cfg, model_cfg, train_cfg):
     normalizer = compute_normalization_stats(full_dataset)
     normalizer.save("data/processed/normalizer.pt")
 
-    # Apply normalizer; precompute Data objects (skipped for torchMD — built on-the-fly)
+    # Apply normalizer; precompute and cache Data objects for faster epoch iteration
     full_dataset.normalizer = normalizer
-    if not use_torchmd:
-        full_dataset.precompute_graphs()
+    full_dataset.precompute_graphs()
```
Comment (RaulPPelaez) [inline scripts/train.py:70]: What does this do?
Are you computing neighbor graphs for the whole dataset before hand? I can see that working for training, it is actually a pretty nifty idea. However, I do not think neighbor construction is going to be a bottleneck in general

File: scripts/visualize.py
Comment (RaulPPelaez) [inline scripts/visualize.py:1]: Consider adding a Jupyter notebook with previously run examples and visualizations

File: src/evaluation/rollout.py [hunk 3]
```diff
@@ -105,7 +132,8 @@ def evaluate_rollout(
         pred_positions[n + 1] = pred_positions[n] + pred_disp
         pred_positions[n + 1] = enforce_boundaries(pred_positions[n + 1], geometry_key, config)
 
-        prev_velocity = torch.tensor(pred_disp / dt, dtype=torch.float32, device=device)
+        if not use_torchmd:
+            prev_velocity = torch.tensor(pred_disp / dt, dtype=torch.float32, device=device)
```
Comment (RaulPPelaez) [inline src/evaluation/rollout.py:136]: Remember that there is no notion of "memory" in your system, so the previous "step" is not even a thing.

File: src/evaluation/rollout.py [hunk 4]
```diff
@@ -73,27 +85,42 @@ def evaluate_rollout(
...
-        edge_index, edge_attr = build_graph(pos_t, graph_method, k_neighbors, cutoff_radius)
-
-        data = Data(
-            x=node_features,
-            edge_index=edge_index,
-            edge_attr=edge_attr,
-            pos=pos_t,
-        ).to(device)
-        data.batch = torch.zeros(N, dtype=torch.long, device=device)
+            # Legacy path: features = [pos, forces, prev_velocity, theta], build graph
```
Comment (RaulPPelaez) [inline src/evaluation/rollout.py:104]: You cannot have a legacy path in such an incipient project. I understand agents love to have these kind of paths. You just have to pester them about it

File: src/evaluation/rollout.py [hunk 5]
```diff
@@ -73,27 +85,42 @@ def evaluate_rollout(
         pos_t = torch.tensor(pos_np, dtype=torch.float32, device=device)
         forces_t = torch.tensor(forces_np, dtype=torch.float32, device=device)
 
-        # Build features
-        if normalizer:
-            pos_norm = normalizer.transform(pos_t.cpu(), "positions").to(device)
-            forces_norm = normalizer.transform(forces_t.cpu(), "forces").to(device)
-            prev_vel_norm = normalizer.transform(prev_velocity.cpu(), "velocities").to(device)
+        if use_torchmd:
```
Comment (RaulPPelaez) [inline src/evaluation/rollout.py:88]: This signals that there is a "not use torchmd" mode, what does that do?

**Review by RaulPPelaez — COMMENTED** [review]
I added some questions

File: scripts/evaluate.py (hunk 1 above)
Comment (IsmaPi) [reply scripts/evaluate.py:163]: This normalizer is basically a normalizer applied to the input features (forces, theta) and output features (displacements) that I am computing during training on the training set and saved there to be able to use it at evaluation time to give the model inputs in the same scale as the training data and also to denormalize the outputs to physical units.

File: scripts/train.py (hunk 2 above)
Comment (IsmaPi) [reply scripts/train.py:70]: The name is a bit misleading, the idea is not to actually compute neighbor graphs since TorchMD-NET already builds its own neighbor list internally at forward time. What it actually caches are the Data objects (normalized forces, theta expansion, positions), so they don't get rebuilt on every call. TLDR: The name is not 100% accurate, and I'll rename it to smth like precompute_data().

File: src/evaluation/rollout.py (hunk 5 above)
Comment (IsmaPi) [reply src/evaluation/rollout.py:88]: Yeah, it's because previously I was training Torch-MD NET models alongside non-Torch-MD NET models, and I needed the distinction, but I need to clean up the repo a bit.

File: src/evaluation/rollout.py (hunk 4 above)
Comment (IsmaPi) [reply src/evaluation/rollout.py:104]: Ok, will do

File: src/evaluation/rollout.py (hunk 3 above)
Comment (IsmaPi) [reply src/evaluation/rollout.py:136]: Yup, this was for the previous models before we talked about it

File: src/evaluation/rollout.py (hunk 5 above)
Comment (IsmaPi) [reply src/evaluation/rollout.py:88]: .

File: scripts/visualize.py
Comment (IsmaPi) [reply scripts/visualize.py:1]: Yes, sounds like a good idea.

File: scripts/train.py (hunk 2 above)
Comment (IsmaPi) [reply scripts/train.py:70]: .

File: scripts/evaluate.py (hunk 1 above)
Comment (IsmaPi) [reply scripts/evaluate.py:163]: .
